### PR TITLE
Support SRV records for service discovery

### LINKED ISSUE #138: Cluster and Service discovery
**Summary:** 

It would be really nice if we could do some kind of cluster/service discovery. 

**Why do we need this?**

1. It would be really helpful for users if they could get a list of public clusters to choose from.
2. Every cluster may have been deployed in a different way. There may be different domains, ports, TLS/non-TLS connections may or may not be available, etc.

**What is already there? What do you see now?**

- Clusters are currently not registered, but there are ideas/plans to register clusters in the Identity Server: #143
- We currently have default ports, and default configuration.

**What is missing? What do you want to see?**

Assuming that clusters will be registered in the Identity Server, we either need to put service information in there, or we need a different mechanism to discover service information.

**How do you propose to implement this?**

In https://github.com/TheThingsIndustries/lorawan-stack/issues/1131 I suggested to use DNS SRV records for service discovery.

A TTN cluster `eu-west.thethings.network` could be discovered through `SRV` records in DNS:

| **Field** | **Value** |
| --------- | --------- |
| _service._proto.name. | `_ttn-v3-{gs,ns,as,js}-{grpc,http,mqtt}._{tcp,tls}.eu-west.thethings.network.` |
| TTL | `<ttl>` |
| class | `IN` |
| priority | `<priority>` |
| weight | `<weight>` |
| port | `8884`/`1884`/`443`/`80`/`8883`/`1883`/... |
| target | domain |

```
_ttn-v3-gs-grpc._tls.eu-west.thethings.network. <ttl> IN <priority> <weight> 8884 eu-west.thethings.network.
_ttn-v3-ns-grpc._tls.eu-west.thethings.network. <ttl> IN <priority> <weight> 8884 eu-west.thethings.network.
_ttn-v3-as-grpc._tls.eu-west.thethings.network. <ttl> IN <priority> <weight> 8884 eu-west.thethings.network.
_ttn-v3-js-grpc._tls.eu-west.thethings.network. <ttl> IN <priority> <weight> 8884 eu-west.thethings.network.
_ttn-v3-gs-grpc._tcp.eu-west.thethings.network. <ttl> IN <priority> <weight> 1884 eu-west.thethings.network.
_ttn-v3-ns-grpc._tcp.eu-west.thethings.network. <ttl> IN <priority> <weight> 1884 eu-west.thethings.network.
_ttn-v3-as-grpc._tcp.eu-west.thethings.network. <ttl> IN <priority> <weight> 1884 eu-west.thethings.network.
_ttn-v3-js-grpc._tcp.eu-west.thethings.network. <ttl> IN <priority> <weight> 1884 eu-west.thethings.network.
...
```

Note that in this example, the SRV records only indicate a port mapping for the services exposed by the cluster (so these records will only have to be set once for a deployment). The target `eu-west.thethings.network` is assumed to be a load balancer in front of the cluster. Alternatively we could use the SRV records for load balancing.
 
We may have to expose some extra information if the GS has multiple UDP endpoints that may have different behavior (different frequency plan for example). This can be discussed later, as I think this is a pretty advanced use case.

**What can you do yourself and what do you need help with?**

Let's first think about cluster registration in the Identity Server: #143

### PR DESCRIPTION
**Pull request by johanstokking**
#### Summary

Closes #138
Replaces #2762

#### Changes

Use SRV records as proposed in #138 via the `pkg/rpcmiddleware/discovery.WithDialer()` gRPC dial option.

Discovery for non-gRPC services is not implemented yet, but we can do that on a case-by-case basis, and consider doing that in specific client Go modules.

#### Testing

Unit testing and local test environment with production DNS (`_ttn-v3-dcs-grpc._tcp.0.0.0.0.0.0.0.D.E.7.5.D.3.B.0.7.join.thethings.industries`)

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/rpcmiddleware/discover/discover.go [hunk 1]
```diff
@@ -77,64 +99,106 @@ func DefaultURL(target string, port int, tls bool) (string, error) {
...
-	return grpc.DialContext(ctx, target, allOpts...)
+
+	return grpc.WithContextDialer(func(ctx context.Context, target string) (net.Conn, error) {
+		var addresses []string
+		if _, _, err := net.SplitHostPort(target); err == nil {
+			addresses = []string{target}
+		} else {
+			if service, ok := ServiceName(role); ok {
+				_, addrs, err := options.dnsResolver.LookupSRV(ctx, service, "tcp", target)
+				if err == nil {
```
Comment (htdvisser) [inline pkg/rpcmiddleware/discover/discover.go:172]: Shouldn't we return lookup errors (other than not found)?

```go
var dnsErr *net.DNSError
if errors.As(err, &dnsErr) {
	if !dnsErr.IsNotFound {
		return nil, err
	}
}
```

Comment (johanstokking) [reply pkg/rpcmiddleware/discover/discover.go:178]: 84617e29b

File: pkg/rpcmiddleware/discover/discover.go [hunk 2]
```diff
@@ -21,9 +21,11 @@ import (
 	"net"
 	"strings"
```
Comment (htdvisser) [inline pkg/rpcmiddleware/discover/discover.go:23]: Nit: We don't usually put empty lines between stdlib imports.

Comment (johanstokking) [reply pkg/rpcmiddleware/discover/discover.go:23]: `gofumports` does that

**Pull request merged**